File: ros2_ws/src/my_cpp_pkg/src/my_first_node.cpp
#include "rclcpp/rclcpp.hpp"

// Class Declaration
class MyNode: public rclcpp::Node   //This declares a class named MyNode that inherits from rclcpp::Node
{
    public:
    MyNode(): Node("cpp_test"), counter_(0)     // Is the constructor for the MyNode class
    { 
        // Logs an informational message "Hello Cpp Node" using the ROS 2 logging mechanism
        RCLCPP_INFO(this->get_logger(), "Hello Cpp Node"); 

        // Sets up a timer to call timerCallback every second.
        timer_ = this->create_wall_timer(std::chrono::seconds(1),
                                        std::bind(&MyNode::timerCallback, this));
    }

    private:
        // Increments and logs a counter value in the timerCallback method
        void timerCallback(){
            counter_++;
            RCLCPP_INFO(this->get_logger(), "Hello %d", counter_);
        }

        rclcpp::TimerBase::SharedPtr timer_;    //A shared pointer to the timer object.
        int counter_;        
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);   //Initialize the node and ROS2 communication
    //Creates a node with shared pointer inside the executable
    auto node = std::make_shared<MyNode>();
    rclcpp::spin(node); //Pause, give spin (useful with callbacks)
    rclcpp::shutdown(); //Shutdown the ROS2 communication
    return 0;
}


/* Notes
        class MyNode: public rclcpp::Node 
        MyNode(): Node("cpp_test"), counter_(0)
    It initializes the base class (rclcpp::Node) with the name "cpp_test". 
    This means that the ROS node will be named "cpp_test".
    It also initializes counter_ to 0.
        rclcpp::init(argc, argv); 
    It takes in two arguments: argc, which stands for "argument count" 
    and represents the number of command-line arguments passed into our program, 
    and argv, which stands for "argument vector" 
    and contains an array of strings representing each argument passed in.
        auto node = std::make_shared<rclcpp::Node>("cpp_test"); 
    Next, we create a shared pointer called "node" which points to an instance of rclcpp::Node.
    This creates a new node with the name "cpp_test".
*/

File: ros2_ws/src/my_py_pkg/my_py_pkg/my_first_node.py
#!/usr/bin/env python3  

import rclpy
from rclpy.node import Node

class MyNode(Node):

    def __init__(self):
        super().__init__("py_test")             #Give name to the node
        self.counter_ = 0
        self.get_logger().info("Hello ROS2")    #Logs a message with info
        self.create_timer(0.5, self.timer_callback)

    def timer_callback(self):
        self.counter_ += 1
        self.get_logger().info("Hello " + str(self.counter_))

def main(args=None):
    rclpy.init(args=args)   #Initialize the node and ROS2 communication
    node = MyNode()         #Give name to the node
    rclpy.spin(node)        #Pause, give spin (useful with callbacks)
    rclpy.shutdown()        #Shutdown the ROS2 communication

#Basic python program structure
if __name__=="__main__":
    main()

""" Notes
- The node is not the executable nor the file
- The name of the node is not the name of the file
- The node is created inside the file

shebang -->(#+!) 
El shebang es una línea que se encuentra al principio de un archivo de script 
y que especifica qué programa debe llamarse para ejecutar el script.
Execute with a Python interpreter, using the env program search path to find it
"""